app/src/main/java/com/example/chatbotandroid/controller/MainActivity.kt

```diff
@@ -0,0 +1,108 @@
+package com.example.chatbotandroid.controller
+
+import androidx.appcompat.app.AppCompatActivity
+import android.os.Bundle
+import com.example.chatbotandroid.R
+import androidx.recyclerview.widget.LinearLayoutManager
+import com.example.chatbotandroid.model.ChatAdapter
+import com.example.chatbotandroid.model.Message
+import com.example.chatbotandroid.utils.Constants
+import com.example.chatbotandroid.utils.Time
+import org.java_websocket.client.WebSocketClient
+import kotlinx.android.synthetic.main.activity_main.*
+import kotlinx.coroutines.*
+import org.java_websocket.handshake.ServerHandshake
+import java.lang.Exception
+import java.net.URI
+
+class MainActivity : AppCompatActivity() {
+    private lateinit var adapter: ChatAdapter
+    private lateinit var webSocketClient: WebSocketClient
+    private val webSocketUri: URI? = URI(Constants.WEB_SOCKET_URL)
+
+    //Life-cycle methods
```

Comment:
one fun thing you may have noticed already with lifecycles on Android is that the UI gets blown away whenever something called a "configuration change" happens, i.e. rotating the screen orientation from portrait to landscape, changing the system theme from light to dark, changing system text size, etc. for example, if we type out and send a few messages, then rotate the screen, all the messages get cleared from the UI.

there's a class called `ViewModel` that helps keep your _data_ alive, in memory, across those configuration changes.

you might then wonder 'why does the UI respond so destructively to configuration changes?' and it's a great question - it's because it gives us android developers an easy way to make an Activity use configuration-specific UI. e.g. if you had a unique landscape layout vs. portrait, the UI would update automagically after a config change, without having a bunch of state to manage (e.g. we wouldn't have to handle unloading the portrait layout and loading the landscape layout when rotating the screen from portrait to landscape)

Reply:
This is super interesting! Thanks for the insight on this :)